library/core/src/ptr/non_null.rs

```diff
@@ -1803,4 +1809,66 @@ mod verify {
         let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    #[kani::proof_for_contract(NonNull::len)]
+    pub fn non_null_check_len() {
+        const SIZE: usize = 20000;
+        // Create a non-deterministic array of size SIZE
+        let arr: [i8; SIZE] = kani::any();  
+        // Get a raw pointer to the array
+        let raw_ptr: *const i8 = arr.as_ptr();  
+        // Create a NonNull slice from the raw pointer
+        let non_null_slice: NonNull<[i8]> = unsafe { NonNull::slice_from_raw_parts(NonNull::new(raw_ptr as *mut i8).unwrap(), SIZE) };  
```

Comment:
Length is a safe operation, so you should be able to generate a non-deterministic `NonNull`. I would suggest you to do the following:
1. Create an implementation of `kani::Arbitrary` for `NonNull<T>`.
2. Use it in your harness like:
```rust
let non_null_ptr: NonNull<i8> = kani::any();
let non_null_slice: NonNull<[i8]> = unsafe { NonNull::slice_from_raw_parts(non_null_ptr, kani::any()) };
```

Reply:
I tried to modify based on your suggestion but got the following error:

```
 |
1797 |             let ptr: *mut T = kani::any();
     |                               ^^^^^^^^^^^ the trait `kani::Arbitrary` is not implemented for `*mut T`
     |
     = help: the following other types implement trait `kani::Arbitrary`:
               ()
               (A, B)
               (A, B, C)
               (A, B, C, D)
               (A, B, C, D, E)
               (A, B, C, D, E, F)
               (A, B, C, D, E, F, G)
               (A, B, C, D, E, F, G, H)
             and 45 others
note: required by a bound in `kani::any`
    --> /Users/jimmy/Desktop/verify-rust-std/library/core/src/lib.rs:426:1
     |
426  | kani_core::kani_lib!(core);
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `any`
     = note: this error originates in the macro `kani_core::kani_intrinsics` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
``` 

```rust
impl<T> kani::Arbitrary for NonNull<T> {
        fn any() -> Self {
            let ptr: *mut T = kani::any();
            NonNull::new(ptr).expect("Non-null pointer expected")
        }
    }
``` 

```rust
#[kani::proof_for_contract(NonNull::len)]
    pub fn non_null_check_len() {
        // Create a non-deterministic NonNull pointer using kani::any()
        let non_null_ptr: NonNull<i8> = kani::any();
        // Create a non-deterministic size using kani::any()
        let size: usize = kani::any();
        // Create a NonNull slice from the non-null pointer and size
        let non_null_slice: NonNull<[i8]> = unsafe { NonNull::slice_from_raw_parts(non_null_ptr, size) };

        // Perform the length check
        let len = non_null_slice.len();
        assert!(len == size);
    }
```

Reply:
You would need to do:
```rust
            let ptr: *mut T = kani::any::<usize>() as *mut T;
```